### PR TITLE
Adding support of Openshift Objects (deploymentconfigs)

### PR DESCRIPTION
**Pull request by andreas131989**
This PR is adding support of scaling deploymentconfig objects in the same way as the other Kubernetes objects are treated. 

Changes:
- Previously the operator deployment manifest and dockerifle used a default nonroot user of Kubernetes. Now, this has changed to use nonroot user in also Openshift
- Added an example deploymentconfig with the random generator
- Identical RBAC for deploymentconfig objects
- A dedicated watch controller for each of the objects (currently supported deployments and deploymentconfigs)
- Ability to natively distinguish between Openshift and Kubernetes by querying the API server on the lookout for openshift objects (specifically in apps.openshift.io/v1 group). The deploymentconfig watcher will be activated only in case of identified openshift resource. 
- Restructured resources to their own packages (deployments and deploymentconfigs). That way it'll be easier to add support of more resources

After we are ok with the structure, design and implementation, next step is to create proper tests for all that and start going towards e2e testing before we add more core functionality.

### CODE REVIEW

**Review by PathogenDE — APPROVED** [review]
Works, and nice arch changes
/lgtm

File: internal/reconciler/reconciler.go [hunk 1]
```diff
@@ -74,14 +103,43 @@ func ReconcileDeployment(ctx context.Context, _client client.Client, deployment
...
 	return nil
 }
 
+func ReconcileDeploymentConfig(ctx context.Context, _client client.Client, deploymentConfig ocv1.DeploymentConfig, state states.State) error {
+	log := ctrl.Log.
+		WithValues("deploymentConfig", deploymentConfig.Name).
+		WithValues("namespace", deploymentConfig.Namespace)
+	stateReplicas, err := sr.NewStateReplicasFromAnnotations(deploymentConfig.GetAnnotations())
+	if err != nil {
+		ctrl.Log.
```
Comment (alex-glv) [inline internal/reconciler/reconciler.go:119]: uses ctrl.log here, but local log var. further

File: internal/reconciler/reconciler.go [hunk 2]
```diff
@@ -74,14 +103,43 @@ func ReconcileDeployment(ctx context.Context, _client client.Client, deployment
...
+			WithValues("namespace", deploymentConfig.Namespace).
+			Error(err, "Cannot calculate state replicas. Please check deploymentConfig annotations. Continuing.")
+		return err
+	}
+	// Now we have all the state settings, we can set the replicas for the deploymentConfig accordingly
+	stateReplica, err := stateReplicas.GetState(state.Name)
+	if err != nil {
+		// TODO here we should do priority filtering, and go down one level of priority to find the lowest set one.
+		// We will ignore any that are not set
+		log.WithValues("set states", stateReplicas).
```
Comment (alex-glv) [inline internal/reconciler/reconciler.go:130]: ☝️

**Review by alex-glv — APPROVED** [review]
one small nitpick, but LGTM!

File: internal/validations/cluster_validation.go [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+		return false, err
+	}
+
+	kubernetesclient, err := kubernetes.NewForConfig(restConfig)
+	if err != nil {
+		return false, err
+	}
+
+	//We use the discovery client to identify the API resources of a given group and version
+	openshiftObjects, err := kubernetesclient.DiscoveryClient.ServerResourcesForGroupVersion(c.OpenshiftObjectGroup)
```
Comment (alex-glv) [inline internal/validations/cluster_validation.go:34]: I must admit it's quite elegant! 🍵

Comment (andreas131989) [reply internal/validations/cluster_validation.go:34]: I enjoyed that one as well!

File: internal/reconciler/reconciler.go (hunk 1 above)
Comment (andreas131989) [reply internal/reconciler/reconciler.go:119]: Nice attention to detail. Good catch!

**Pull request merged**